// built.io-browserify/lib/extension.js
var R 				      			= require('ramda');
var utility 		      		= require('./utilities/utility');
/*
	Constants
*/
var extensionPath = '/functions'; 
/**
  @class Extension
  @classdesc   Built.Extension allows you to add your own business logic to built.io. The business logic is written in JavaScript, and executed on the cloud.
*/

/**
Use the function to execute a function deployed on built.io Extension.
@function execute
@param fnName {Function} The name of the function that needs to be executed
@param reqBody {object} The request body that needs to be send with the request (Optional)
@param reqHeaders {object} Headers that need to be sent along with the request (Optional)
@instance
@memberof Extension
@example
  var app = Built.App('api_key');
  app.Extension.execute('submit')
    .then(function(result){
      // do somethings
    });
@return {Promise<Variable>}
*/
module.exports.execute = R.curry(function(app, fnName, reqBody, reqHeaders){
	var adaptor 		  = app.options.adaptor; 
	var url 					= getURL(fnName,app);
  var requestObject = utility.getAdaptorObj('POST',url,getCombinedHeaders(app,reqHeaders),reqBody,null);
  return adaptor.makeCall(requestObject).then(function(response) {
  	return R.prop('result', response.entity);
  });
});

function getURL (fnName,app) {
	return app.getURL() + extensionPath + '/' + fnName;
}

function getCombinedHeaders(app,reqHeaders){
	reqHeaders = reqHeaders || {};
	var newHeaders = R.mixin(app.getHeaders(),reqHeaders);
	return newHeaders;
}
