// built.io-browserify/lib/role.js
var R                     = require('ramda');
var utility               = require('./utilities/utility');
var instanceMethodBuilder = require('./utilities/instanceMethodBuilder')();
var objectCons            = require('./object');
var constants             = require('./constants');


/**
  @class Role
  @classdesc
    A Role is logical grouping of a set of users. Role inherts from {@link Object}.
  @description
    Constructs a role
  @param uid Uid of a role on built.io
  @example
    var role = Built.App('api_key').Role('uid');
  @return {Role}
*/
var roleCons = module.exports = R.curry(function(app,cls,headers,query,data,delta) {
  var returnObj  = objectCons(app,cls,headers,{},data,delta); 
  return instanceMethodBuilder.build(module.exports,returnObj);
});
utility.copyProperties(module.exports, objectCons);

/**
  Returns a query instance for Role Class.
  @function getRoleQuery
  @instance
  @memberof Role
  @example
    var Role  = Built.App('api_key').Role;
    var query = Role.getRoleQuery();
    query
    .where('name','admin')
    .exec()
      .then(function(roles){
  
      });

  @return {Query}
*/
module.exports.getRoleQuery = function(app){
  return function(){
    return app.Class(constants.APP_ROLE_CLS).Query();
  }
};

var addHelper = R.curry(function (parameter,uid,role){
  var newDelta        = getMixinedDelta(role);
  var newData         = getMixinedData(role)
  var newRoles        = getArrayByParameter(parameter,role);
  newRoles.push(uid);
  newDelta[parameter] = newRoles;
  newData[parameter]  = newRoles;
  return role.cls.ObjectCons(role.getHeaders(),{},newData,newDelta);
});

/**
  Adds a new Role
  @function addRole
  @param {String} role_uid The uid of the role to be added
  @instance
  @example
    var role  = Built.App('api_key').Role();
    role      = role
                .setName('developers')
                .addRole('abc123');
  @memberof Role
  @return  {Role}
*/
module.exports.addRole = addHelper('roles');
instanceMethodBuilder.define('addRole',2);

/**
  Adds a new user to this role
  @function addUser
  @param {String} user_uid The uid of the user to be added
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role();
    role      = role.addUser('abc123');
  @return  {Role}
*/
module.exports.addUser = addHelper('users');
instanceMethodBuilder.define('addUser',2);

/**
  Gets the name of the role
  @function getName
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(function(role){
        role.getName();
      })
  @return  {String}
*/
module.exports.getName = module.exports.get('name');
instanceMethodBuilder.define('getName',1);

/**
  Gets the subroles contained in this role. Subroles are a subset of Role.
  @function getRoles
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(function(role){
        role.getRoles();
      });
  @return  {Array}
*/
module.exports.getRoles = module.exports.get('roles');
instanceMethodBuilder.define('getRoles',1);

/**
  Get the uid of the role
  @function getRoleUid
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role('abc123');
    var uid   = role.getRoleUid(); 
  @return  {String}
*/
module.exports.getRoleUid = module.exports.get('uid');
instanceMethodBuilder.define('getRoleUid',1);

/**
  Gets the users belonging to this role
  @function getUsers
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(function(role){
        role.getUsers();
      })
  @return  {Array}
*/
module.exports.getUsers = module.exports.get('users');
instanceMethodBuilder.define('getUsers',1);

/**
  Assigns a name to this role
  @function setName
  @param {String} name Name to be assigned to this role
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role();
    role      = role.setName('developers');
  @return  {Role}
*/
module.exports.setName = module.exports.set('name');
instanceMethodBuilder.define('setName',2);
/**
  Assigns a uid to this role
  @function setRoleUid
  @param {String} uid Uid to be assigned to this role
  @instance
  @memberof Role
  @example
    var role  = Built.App('api_key').Role();
    role      = role.setRoleUid('abc123');
  @return  {Role}
*/
module.exports.setRoleUid = module.exports.set('uid');
instanceMethodBuilder.define('setRoleUid',2);

var hashelper = R.curry(function(type,uid,role){
  return R.contains(uid,getData(role)[type]);
});

/**
  Checks whether a sub role exists in this role
  @function hasRole
  @param {String} role_uid Uid of the role whose existence needs to be checked
  @instance
  @memberof Role
  @example
    var role    = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(fuction(role){
        role.hasRole('pqr123'); // pass sub role uid
      });
  @return  {Boolean}
*/
module.exports.hasRole = hashelper('roles');
instanceMethodBuilder.define('hasRole',2);
/**
  Checks whether a user belongs to this role
  @function hasUser
  @param {String} user_uid Uid of the user whose existence needs to be checked
  @instance
  @memberof Role
  @example
    var role    = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(fuction(role){
        role.hasUser('pqr123'); // pass user uid
      });
  @return  {Boolean}
*/
module.exports.hasUser = hashelper('users');
instanceMethodBuilder.define('hasUser',2);

var removeHelper = R.curry(function(type,uid,role){
  var newDelta = getMixinedDelta(role);
  var newData  = getMixinedData(role);
  var array    = getArrayByParameter(type,role);
  var index    = array.indexOf(uid);
  if(index > -1)
    array.splice(index,1);
  newDelta[type] = array;
  newData[type]  = array;  
  return role.cls.ObjectCons(role.getHeaders(),{},newData,newDelta);  
});

/**
  Removes a sub role from this role
  @function removeRole
  @param {String} role_uid Uid of the sub role which should be removed from this role
  @instance
  @memberof Role
  @example
    var role = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(function(role){
        if(role.hasRole('pqr123'));
          return role.removeRole('pqr123'); //pass role uid
        else
          return role;
      });
  @return  {Role}
*/
module.exports.removeRole = removeHelper('roles');
instanceMethodBuilder.define('removeRole',2);

/**
  Removes a user from this role
  @function removeUser
  @param {String} user_uid Uid of the user who should be removed from this role
  @example
    var role = Built.App('api_key').Role('abc123');
    role.fetch()
      .then(function(role){
        if(role.hasUser('pqr123'));
          return role.removeUser('pqr123'); //pass user uid
        else
          return role;
      });
  @instance
  @memberof Role
  @return  {Role}
*/
module.exports.removeUser = removeHelper('users');
instanceMethodBuilder.define('removeUser',2);

function getMixinedDelta(role){
 return R.mixin({},getDelta(role));
};

function getMixinedData(role){
 return R.mixin({},getData(role));
};

function getDelta(role){
  return role.delta;
}

function getData(role){
  return role.data;
}

/*
This function checks whether array(roles|users) exists in role 
if it does makes a new array containing the same contents or retunrs a new array
*/
function getArrayByParameter(parameter,role){
  var array;
  if(getData(role)[parameter])
    var array  = [].concat(getData(role)[parameter]);
  else
    array  = [];
  return array;
}
